libraries/AP_Scripting/lua_repl.cpp

```diff
@@ -104,26 +104,41 @@ int lua_scripts::pushline(lua_State *L, int firstline) {
     terminal_print(get_prompt(L, firstline));
 
     while (terminal.session) {
-        // reseek to where we need input from, as invalid reads could have done weird stuff, and we want to start from the last valid input
-        int input_fd = AP::FS().open(REPL_IN, O_RDONLY);
-        if (input_fd != -1) {
-            AP::FS().lseek(input_fd, terminal.input_offset, SEEK_SET);
-            ssize_t read_bytes = AP::FS().read(input_fd, buffer, ARRAY_SIZE(buffer) - 1);
-            AP::FS().close(input_fd);
-            if (read_bytes > 0) {
-                // locate the first newline
-                char * newline_chr = strchr(buffer, '\n');
-                if (newline_chr != NULL) {
-                    newline_chr[0] = '\0';
-                    // only advance to the newline
-                    l = strlen(buffer);
-                    terminal.input_offset += l + 1;
+        if (terminal.type == AP_Scripting::relp_type::REPL_FILE)  {
+            // reseek to where we need input from, as invalid reads could have done weird stuff, and we want to start from the last valid input
+            int input_fd = AP::FS().open(REPL_IN, O_RDONLY);
+            if (input_fd != -1) {
+                AP::FS().lseek(input_fd, terminal.input_offset, SEEK_SET);
+                ssize_t read_bytes = AP::FS().read(input_fd, buffer, ARRAY_SIZE(buffer) - 1);
+                AP::FS().close(input_fd);
+                if (read_bytes > 0) {
+                    // locate the first newline
+                    char * newline_chr = strchr(buffer, '\n');
+                    if (newline_chr != NULL) {
+                        newline_chr[0] = '\0';
+                        // only advance to the newline
+                        l = strlen(buffer);
+                        terminal.input_offset += l + 1;
+                        break;
+                    }
+                }
+            }
+            // wait for any input
+            hal.scheduler->delay(100);
+        } else {
+            // can only read a byte at a time as we have to look out for the newline
+            if (terminal.uart->available()) {
+                buffer[l] = terminal.uart->read();
+                if (buffer[l] == '\n') {
+                    buffer[l + 1] = 0;
                     break;
                 }
+                l++;
+            } else {
+                // wait for any input
+                hal.scheduler->delay(100);
```

Comment:
We could remove this delay for a faster update rate, any reason why going flat out would be bad?

Reply:
Was basically to not eat CPU time we didn't need. We're currently the lowest priority so it shouldn't be to bad, but that doesn't stop us from having other odd effects like eating memory bandwidth when we don't need to. For the UART case cutting this down to ~5 ms is probably easily fine. Even the file case could probably come down a bit. I think I was more worried about scripting being at IO priority rather then a lower priority at a time, so I was probably a bit cautious.

Reply:
changed to 5ms